### PR TITLE
8310944: [Lilliput/JDK17] Streamline and cleanup GC code

### PR DESCRIPTION
**Pull request by rkennke**
We have a bunch of ugly hacks in GC code paths where we do stuff like:
if (UseCompactObjectHeaders) {
 // fetch object Klass* or size in a safe way...
} else {
  // fetch object Klass* or size in the traditional way
}

This happens in various places in similar ways. I propose to add relevant methods to oopDesc, and make all GC code use those new methods instead.

This helps streamline and clean up a bunch of GC code, and bring it closer to the JDK22 upstreaming PR.

Testing:
 - [x] hotspot_gc
 - [x] hotspot_gc (+UCOH)
 - [x] tier1
 - [x] tier1 (+ShenandoahGC)
 - [x] tier1 (+ParallelGC)
 - [x] tier1 (+SerialGC)
 - [x] tier1 (+UCOH)
 - [x] tier1 (+UCOH +ShenandoahGC)
 - [x] tier1 (+UCOH +ParallelGC)
 - [x] tier1 (+UCOH +SerialGC)
 - [x] tier2
 - [x] tier2 (+UCOH)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310944](https://bugs.openjdk.org/browse/JDK-8310944): [Lilliput/JDK17] Streamline and cleanup GC code (**Bug** - P4)

### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/lilliput-jdk17u.git pull/50/head:pull/50` \
`$ git checkout pull/50`

Update a local copy of the PR: \
`$ git checkout pull/50` \
`$ git pull https://git.openjdk.org/lilliput-jdk17u.git pull/50/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 50`

View PR using the GUI difftool: \
`$ git pr show -t 50`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/lilliput-jdk17u/pull/50.diff">https://git.openjdk.org/lilliput-jdk17u/pull/50.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/lilliput-jdk17u/pull/50#issuecomment-1612778394)

### CODE REVIEW

**Review by shipilev — COMMENTED** [review]
I have questions :)

File: src/hotspot/share/oops/oop.inline.hpp [hunk 1]
```diff
@@ -249,6 +257,53 @@ int oopDesc::size_given_klass(Klass* klass)  {
...
+
+size_t oopDesc::forward_safe_size() {
+  return size_given_klass(forward_safe_klass());
+}
+
+void oopDesc::forward_safe_init_mark() {
+  if (UseCompactObjectHeaders) {
+    set_mark(forward_safe_klass()->prototype_header());
+  } else {
+    set_mark(markWord::prototype());
```
[the remark this answers is not in the document]
Comment (rkennke) [reply src/hotspot/share/oops/oop.inline.hpp:303]: Right. In-fact, this means we can just fix init_mark() and use that.

File: src/hotspot/share/oops/oop.inline.hpp [hunk 2]
```diff
@@ -92,6 +92,14 @@ markWord oopDesc::resolve_mark() const {
   return hdr;
 }
 
+markWord oopDesc::prototype_mark() const {
+  if (UseCompactObjectHeaders) {
+    return klass()->prototype_header();
```
[the remark this answers is not in the document]
Comment (rkennke) [reply src/hotspot/share/oops/oop.inline.hpp:97]: I don't think that any objects would be forwarded during serial mark-sweep marking phase.

File: src/hotspot/share/oops/oop.inline.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkennke) [reply src/hotspot/share/oops/oop.inline.hpp:303]: Actually, no. We need to resolve the forwardee.

File: src/hotspot/share/oops/oop.inline.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (shipilev) [reply src/hotspot/share/oops/oop.inline.hpp:97]: Right, okay.

**Review by shipilev — APPROVED** [review]
OK!

**Pull request closed without merging**